hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngrade.java

```diff
@@ -90,14 +90,34 @@ public static boolean needsUpgradeOrDowngrade(HoodieTableMetaClient metaClient,
 
   public static boolean needsUpgrade(HoodieTableMetaClient metaClient, HoodieWriteConfig config, HoodieTableVersion toWriteVersion) {
     HoodieTableVersion fromTableVersion = metaClient.getTableConfig().getTableVersion();
-    // If table version is less than SIX, then we need to upgrade to SIX first before upgrading to any other version, irrespective of autoUpgrade flag
+    if (fromTableVersion.versionCode() >= toWriteVersion.versionCode()) {
+      // if the table version is greater than or equal to the write version, then this is not an upgrade
+      LOG.warn("Table version {} is greater than or equal to write version {}. No upgrade needed", fromTableVersion, toWriteVersion);
+      return false;
+    }
     if (fromTableVersion.versionCode() < HoodieTableVersion.SIX.versionCode() && toWriteVersion.versionCode() >= HoodieTableVersion.EIGHT.versionCode()) {
```

Comment:
As discussed, let's enforce that the fromVersion should be at least SIX.  Disable any tests (e.g., 4 -> 5, 5 -> 6) that are not applicable any more (we can have a separate PR to clean up the code).
```suggestion
    if (fromTableVersion.versionCode() < HoodieTableVersion.SIX.versionCode())
```

Reply:
Ack will do so based on discussion.